### ISSUE TITLE
Podcast endpoints?

### ISSUE THREAD

**Issue by seahorsepip** [spotify/web-api#551]
Add endpoints to get, save, remove etc. podcasts.

**Comment by hughrawlinson** [comment]
Thanks for the feature request @seahorsepip!

**Comment by HenrikFricke** [comment]
@hughrawlinson Can you give us more insights? Is it on a roadmap? Thanks :)

**Comment by hughrawlinson** [comment]
Hi @HenrikFricke - we don't have a public roadmap at the moment, but any updates will be posted on developer.spotify.com, and this thread will be updated. Right now, there's no timeline for podcast endpoints.

**Comment by HenrikFricke** [comment]
@hughrawlinson okay, thanks for the answer :)

**Comment by ridhwaans** [comment]
+1 asking to bring podcasts to the web api. My pain is I cant fetch metadata for podcasts like this one https://open.spotify.com/episode/5TSsSAR7MHV2aC7jVvpvBw?si=ZMg7hkWOQa6lI383S2QEzQ  
I'll have to start looking at other services

**Comment by RWOverdijk** [comment]
I'd love this as a feature, too. Just adding this due to radio silence

**Comment by jamescridland** [comment]
+1 from me - I'd like to link to Spotify from podcast pages [like this one](https://podnews.net/podcast/201671138); as the #2 podcast destination in the world, it would be brilliant to link further to Spotify.

**Comment by kapilbahadur** [comment]
+1 Need this API to be open. There is no in-app feature to view the podcasts I've seen unless I individually go to a podcast, and filter by progress.

**Comment by gilibaus** [comment]
+1 from me too

**Comment by captainbenno** [comment]
@hughrawlinson who do we have to bribe to get this functionality available?

**Comment by patalanov** [comment]
+ 1please

**Comment by Jakubem** [comment]
+1, that would be a great feature

**Comment by Worie** [comment]
indeed, +1

**Comment by emily-morris** [comment]
+1

**Comment by Grace** [comment]
+1

**Comment by suissemaxx** [comment]
+1

**Comment by rochabianca** [comment]
+1

**Comment by nathangathright** [comment]
Podcasts now appear to be showing up in search on [open.spotify.com](https://open.spotify.com/search/podcasts/dissect)

**Comment by jamescridland** [comment]
@nathangathright That's nice to know - but sadly, there's still no sign of them in the [web API](https://developer.spotify.com/documentation/web-api/reference/search/search/). Unless we know different!

**Comment by jamescridland** [comment]
Actually... ah, this is interesting. So, podcasts are appearing under this URL:

https://api.spotify.com/v1/search?type=album%2Cartist%2Cplaylist%2Ctrack%2Cshow_audio%2Cepisode_audio&q=podnews*&decorate_restrictions=false&best_match=true&limit=50&userless=true&market=AU

This works ONLY if you have a valid token. There appears to be an Authorization header sent.

Here's a snippet of what it returns:

````
"shows" : {
    "href" : "https://api.spotify.com/v1/search?query=podne*&type=show&market=AU&offset=0&limit=50",
    "items" : [ {
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IL", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "RO", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN", "ZA" ],
      "copyrights" : [ ],
      "description" : "Weekly updates for the podcast and on-demand industry, from podnews.net. Sign up for daily news via email at https://podnews.net/  Our daily podcast is available in iTunes and other podcast apps. This podcast is hosted by Omny Studio - https://omnystudio.com/learn",
      "explicit" : false,
      "external_urls" : {
        "spotify" : "https://open.spotify.com/show/6p1jX1AC0O7yiZ5JdyOmfp"
      },
      "href" : "https://api.spotify.com/v1/shows/6p1jX1AC0O7yiZ5JdyOmfp",
      "id" : "6p1jX1AC0O7yiZ5JdyOmfp",
      "images" : [ {
        "height" : 640,
        "url" : "https://i.scdn.co/image/8cb8a92e7d9235172b5159500326d6c5cbfdd92e",
        "width" : 640
      }, {
        "height" : 300,
        "url" : "https://i.scdn.co/image/ed13a822c594f9439660c838fe8a67f63618b896",
        "width" : 300
      }, {
        "height" : 64,
        "url" : "https://i.scdn.co/image/d33567933e11b4a1389e31378e50851d9d72ce02",
        "width" : 64
      } ],
      "languages" : [ "en-GB" ],
      "media_type" : "audio",
      "name" : "Podnews - weekly podcast news",
      "publisher" : "Podnews.net",
      "type" : "show",
      "uri" : "spotify:show:6p1jX1AC0O7yiZ5JdyOmfp"
    }
````

...so the "type" is the [undocumented](https://developer.spotify.com/documentation/web-api/reference/search/search/) "show" in this instance.

**Comment by maracuja-juice** [comment]
@jamescridland But that only works if you know what podcast to search for right?

**Comment by jamescridland** [comment]
@maracuja-juice Yes, I haven't tried any of the other API end points. My suspicion is that they should all behave in the same way - use the undocumented "show". But I've not tried them or tested them.

**Comment by drpancake** [comment]
@jamescridland I couldn't find a documented API endpoint to do this when I was searching a few weeks ago - not for lack of trying.

The web player exposes a bunch of undocumented APIs for almost anything you might need, including podcast search, but obtaining a valid token in a repeatable automated way is an adventure.

**Comment by maracuja-juice** [comment]
@drpancake The normal token when using the web api doesn't work for these undocumented endpoints then, right?

**Comment by drpancake** [comment]
@maracuja-juice yeah unless something changed in the intervening few weeks.

**Comment by drjackson9999** [comment]
+1

**Comment by jamescridland** [comment]
I'm successfully querying this undocumented endpoint using a token from the standard web API. It works just fine (though searching for a podcast called "The Daily" is rather more painful than it needs be).

**Comment by drjackson9999** [comment]
Yeah, I'm able to use the search, and also the currently-playing end point as entry points into a specific podcast (though the object structure is very different between "tracks" and "episodes") using the normal bearer token. I've then been able to get information on the "show" and the "episode" using hrefs returned in the json by these initial end points. 

Samples that worked for me, using just a simple get at these two urls, using the normal bearer token:

https://api.spotify.com/v1/shows/2SuODPath8vx2W3JNHluKA
https://api.spotify.com/v1/episodes/2lqNevhLx08QPmvCLjhS4g

Things that haven't worked for me that I'd like to have working in an ideal world:
recently played (works of course, but podcasts aren't includedin the result set)
add to queue (gives me an invalid uri error)
add to playlist (invalid uri error again, not all that surprising given you can't add a podcast to a playlist in any spotify player)
add to library (same as playlist, so not all that surprising)

I just want to listen to Spotify podcasts offline on my new Samsung Galaxy Watch! It's driving me nuts that I can't. I was hoping the APIs would give me a workaround but I've had no luck so far.

**Comment by drjackson9999** [comment]
As a quick additional note, you can search using a type of "episode" to find specific podcasts, in addition to type=show

Simple example: 

https://api.spotify.com/v1/search?q=dan%20carlin&type=episode

vs 

https://api.spotify.com/v1/search?q=dan%20carlin&type=show

or with both:

https://api.spotify.com/v1/search?q=dan%20carlin&type=episode,show

**Comment by jscholes** [comment]
> add to queue (gives me an invalid uri error)

Using which endpoint?

**Comment by drjackson9999** [comment]
Add to queue probably isn't the right term, sorry. Play now would be better. I'm using this one:

https://api.spotify.com/v1/player/play

Documented here:

https://developer.spotify.com/documentation/web-api/reference/player/start-a-users-playback/

**Comment by esl-douglas** [comment]
I can't find **followed** podcasts in any of the endpoints available.

**Comment by esl-douglas** [comment]
Nevermind, just found. 

GET https://api.spotify.com/v1/me/shows

**Comment by ngirot** [comment]
You have two endpoints :
GET https://api.spotify.com/v1/me/shows
GET https://api.spotify.com/v1/me/shows/{showId}

But there is some missing information with the second ; like "resume_point" for each podcast entries. If i use a token generated thought the web client, everything is fine, but not a token generated with Oauth (Authorization Code) with all scopes available.

Do someone knows if there is a way to get this data ?

**Comment by jscholes** [comment]
These undocumented endpoints were created for use with Spotify's own Web Player, not for use in third party projects.  That's why a regular token, as issued to a user of a third party application, is not sufficient to access them.

**Comment by tommienu** [comment]
+1

**Comment by klayveR** [comment]
+1

I assume there's still no way to use episode uris with the https://api.spotify.com/v1/me/player/play endpoint?

**Comment by tonyravioli** [comment]
+1

**Comment by JMdoubleU** [comment]
+1

**Comment by ragnorc** [comment]
+1

**Comment by kylepw** [comment]
+1

**Comment by mikenye** [comment]
Furthermore, it would be extremely helpful if we could access podcast statistics (ie: subscriber count, starts, streams, etc) via API, so that we can pull statistics from different podcast platforms together to get a holistic view of one's statistics across all podcast platforms.

**Comment by romellfudi** [comment]
+1

**Comment by Meandmybadself** [comment]
Searching against this endpoint appears to provide podcast results:
GET https://api.spotify.com/v1/search?type=episode_audio&q=treetop%20barbie&decorate_restrictions=false&limit=5&userless=true&market=US

Note: When making the request, provide the header:
`Authorization: Bearer ACCESS_TOKEN`

**Comment by itsme112358** [comment]
Apparently the endpoint exists as @jamescridland  mentioned, it would be nice if it was documented though. 
First I tried to handle podcasts like artists, then playlists before I found this post. 
It would still be nice to get a documentation. All I want is to get the total number of episodes released. I got that, but it seems like a bit overkill. Since the limit parameter seems to be ignored, I always get 50 entries. Also I can't filter by fields. This means, I get a ~200KB result, just for one int, that's not very efficient :D

**Comment by pauljl1** [comment]
# Undocumented Spotify Podcasts API

```:market``` represents a Spotify recognised [ISO country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).

## Show endpoints

```:show_id``` represents a Spotify ID, e.g. '0SUThziJbWNcuVmlmlsvva'

### Find single, by ID - shows

```
https://api.spotify.com/v1/shows/:show_id?market=:market
```

Example: ```https://api.spotify.com/v1/shows/0SUThziJbWNcuVmlmlsvva?market=:market```

### Find multiple, by ID - shows

```
https://api.spotify.com/v1/shows?ids=:show_id,:show_id&market=:market
```

Example: ```https://api.spotify.com/v1/shows?ids=0SUThziJbWNcuVmlmlsvva,3Cct7ZWmxHNAtLgBTqjC5v&market=:market```

### Retrieving episodes for a show with limits and offset 

This doesn't seem to work at the moment.

```
https://api.spotify.com/v1/shows/:show_id?limit=1&offset=1&market=:market
```

Example: ```https://api.spotify.com/v1/shows/0SUThziJbWNcuVmlmlsvva?limit=1&offset=1&market=:market```

### Search - shows

Find a single show by keywords, following [Search API](https://developer.spotify.com/documentation/web-api/reference-beta/#category-search).

```:query``` represents a search query, e.g. 'The Horne Section Podcast'

```
https://api.spotify.com/v1/search?q=:query&type=show&market=:market
```

Example: ```https://api.spotify.com/v1/search?q=horne+section&type=show&market=:market```

**Note**: market is *not* optional when searching for shows, limit and offset still work as normal.


### Expected JSON

If you find one ID, you will be given the episodes. If you're searching for multiple, episodes will be omitted.

- ```available_markets``` - array (string)
- ```copyrights``` - array (string)
- ```description``` - string
- ```episodes``` - hash (see episodes)
- ```explicit``` - boolean
- ```external_urls``` - hash
- ```href``` - string
- ```id``` - string
- ```images``` - array (hash)
- ```is_externally_hosted``` - boolean
- ```languages``` - array
- ```media_type``` - string
- ```name``` - string
- ```publisher``` - string
- ```publisher``` - string
- ```type``` - string
- ```uri``` - string

## Episode end points

```:episode_id``` represents a Spotify ID, e.g. '6GO8UvfFR1kQmnNdmdffFG'

### Find single, by ID - episodes

```
https://api.spotify.com/v1/shows/:show_id&market=:market
```

Example: ```https://api.spotify.com/v1/episodes/6GO8UvfFR1kQmnNdmdffFG&market=:market```

### Find multiple, by ID - episodes

```
https://api.spotify.com/v1/shows?ids=:show_id,:show_id&market=:market
```

Example: ```https://api.spotify.com/v1/episodes?ids=6GO8UvfFR1kQmnNdmdffFG,5LJ8pq2x1A7DzBCPJvz7pY&market=:market```

### Search - episodes

Find a single episode by keywords, following [Search API](https://developer.spotify.com/documentation/web-api/reference-beta/#category-search).

```:query``` represents a search query, e.g. 'The Horne Section Podcast'

```
https://api.spotify.com/v1/search?q=:query&type=episode&market=:market
```

Example: ```https://api.spotify.com/v1/search?q=horne+section+al+murry&type=show&market=:market```

**Note**: market is *not* optional when searching for episodes, limit and offset still work as normal.

### Expected JSON

If you find one ID, you will be given the show. If you're using the show for the episodes, show will not be included in the JSON for episodes.

- ```audio_preview_url``` - string
- ```description``` - string
- ```duration_ms``` - integer
- ```explicit``` - boolean
- ```external_urls``` - hash
- ```href``` - string
- ```id``` - string
- ```images``` - array (hash)
- ```is_externally_hosted``` - boolean
- ```is_playable``` - boolean
- ```language``` - string
- ```languages``` - array
- ```name``` - string
- ```release_date``` - string
- ```release_date_precision``` - string
- ```release_date_precision``` - string
- ```show``` - hash (see show, only outputted if using episode API endpoints)
- ```type``` - string
- ```uri``` - string

## Caveats

I've been doing some experimental work with this API and have found the following:

- it's not yet possible to add or remove episodes to a playlist using the API
- it IS possible to reorder tracks using the API, however, episodes that were once added and are no longer available are output as an empty item. In other words, the ['track' hash of the 'items' array is missing](https://developer.spotify.com/documentation/web-api/reference-beta/#endpoint-get-playlists-tracks) for episodes that are no longer available; the 'added_at', 'added_by', 'is_local' still appear. 

Therefore, I think most of the show and episode API functionality is more for data retrieval than processing.

**Comment by tudi2d** [comment]
Official endpoints: [Search, browse and follow podcasts using the new Podcast APIs](https://developer.spotify.com/community/news/2020/03/20/introducing-podcasts-api/ )

**Web API Reference:**
- [Episodes](https://developer.spotify.com/documentation/web-api/reference-beta/#category-episodes)
- [Shows](https://developer.spotify.com/documentation/web-api/reference-beta/#category-shows)

**Comment by shafiemukhre** [comment]
I am trying do stream the audio track from the spotify endpoint, is this possible? One of the use cases that I am thinking is to do analysis with this audio track

**Comment by pedronaresi** [comment]
Are the "Adding episodes to a playlist" endpoints working? Been trying for while and not getting sucess.
